src/burnchains/db.rs

```diff
@@ -0,0 +1,351 @@
+        let mut stmt = self.sql_tx.prepare(sql)?;
+        for op in block_ops.iter() {
+            let serialized_op = serde_json::to_string(op)
+                .expect("Failed to serialize parsed BlockstackOp");
+            let args: &[&dyn ToSql] = &[block_hash,
+                                        &serialized_op];
+            stmt.execute(args)?;
+        }
+        Ok(())
+    }
+
+    fn commit(self) -> Result<(), BurnchainError> {
+        self.sql_tx.commit()
+            .map_err(BurnchainError::from)
+    }
+}
+
+impl BurnchainDb {
+    pub fn connect(path: &str, first_block_height: u64, first_burn_header_hash: &BurnchainHeaderHash,
+                   first_burn_header_timestamp: u64, readwrite: bool) -> Result<BurnchainDb, BurnchainError> {
+        let mut create_flag = false;
+        let open_flags = match fs::metadata(path) {
+            Err(e) => {
+                if e.kind() == io::ErrorKind::NotFound {
+                    // need to create 
+                    if readwrite {
+                        create_flag = true;
+                        OpenFlags::SQLITE_OPEN_READ_WRITE | OpenFlags::SQLITE_OPEN_CREATE
+                    } else {
+                        return Err(BurnchainError::from(DBError::NoDBError));
+                    }
+                } else {
+                    return Err(BurnchainError::from(DBError::IOError(e)));
+                }
+            },
+            Ok(_md) => {
+                // can just open 
+                if readwrite {
+                    OpenFlags::SQLITE_OPEN_READ_WRITE
+                }
+                else {
+                    OpenFlags::SQLITE_OPEN_READ_ONLY
+                }
+            }
+        };
+
+        let mut db = BurnchainDb {
+            conn: Connection::open_with_flags(path, open_flags)
+                .expect(&format!("FAILED to open: {}", path))
+        };
+
+        if create_flag {
+            let db_tx = db.start_transaction()?;
+            db_tx.sql_tx.execute_batch(BURNCHAIN_DB_SCHEMA)?;
+
+            let first_block_header = BurnchainBlockHeader {
+                block_height: first_block_height,
+                block_hash: first_burn_header_hash.clone(),
+                timestamp: first_burn_header_timestamp,
+                num_txs: 0,
+                parent_block_hash: BurnchainHeaderHash::sentinel()
+            };
+
+
+            db_tx.store_burnchain_db_entry(&first_block_header)?;
+            db_tx.commit()?;
+        }
+
+        Ok(db)
+    }
+
+    pub fn open(path: &str, readwrite: bool) -> Result<BurnchainDb, BurnchainError> {
+        let open_flags = if readwrite {
+            OpenFlags::SQLITE_OPEN_READ_WRITE
+        } else {
+            OpenFlags::SQLITE_OPEN_READ_ONLY
+        };
+        let conn = Connection::open_with_flags(path, open_flags)?;
+
+        Ok(BurnchainDb { conn })
+    }
+
+    fn start_transaction<'a>(&'a mut self) -> Result<BurnchainDbTransaction<'a>, BurnchainError> {
```

Comment:
Not blocking, but would you be open to changing this method name to `tx_begin()`?  It's the name that's currently used in all other DB classes to instantiate transactions against the DB.

Reply:
Changed in 99ed20971